### ISSUE TITLE
AWS credentials

### ISSUE THREAD

**Issue by michailbrynard** [zalando/spilo#199]
I see there's a GOOGLE_APPLICATION_CREDENTIALS environmental variable. Is there a similar one for AWS? Or how does the S3 bucket authentication work?

**Comment by alexeyklyukin** [comment]
ON AWS we use the instance profiles to avoid passing the credentials around, i.e. WAL-E is called with the `--aws-instance-profile` option. Basically, we use the credentials of the EC2 instance.

**Comment by michailbrynard** [comment]
Thanks for the quick response. It could be useful to also add an option for loading AWS keys via env variables for cases where the Spilo is not running on AWS infrastructure, but one still wants to backup to S3. For example, I'm running on GCloud, but was thinking of backing up to S3. However, for now it seems easiest just to backup to Google Cloud Storage instead.

**Comment by michailbrynard** [comment]
I see in your code, you also have:
```
    elif provider == PROVIDER_GOOGLE and 'WAL_GCS_BUCKET' in placeholders:
        placeholders['USE_WALE'] = True
        placeholders.setdefault('GOOGLE_APPLICATION_CREDENTIALS', '')
```
This seems like I can only have the backups linked to GCloud if I am running the container on Gcloud. Is this by design? What about the case where I want to run the container locally, but backup to a GCloud or AWS bucket?

**Comment by CyberDem0n** [comment]
> This seems like I can only have the backups linked to GCloud if I am running the container on Gcloud. Is this by design? What about the case where I want to run the container locally, but backup to a GCloud or AWS bucket?

That's true, it's not very flexible right now, but we always welcome PR with improvements.

**Comment by michailbrynard** [comment]
Thanks. I'll see if I can get it working and then try to find some time for a PR.

**Comment by michailbrynard** [comment]
@CyberDem0n Currently, Spilo works by checking the 169.254.169.254 endpoint to determine whether the container is running locally or on AWS, Openstack or GCloud. 

It then uses the determined provider to decide the WAL backup destination.

Would it be acceptable to modify the behavior to always use the AWS bucket for backups if WAL_S3_BUCKET is specified and to always use google cloud storage if WAL_GCS_BUCKET is specified, regardless of the provider?

This is what I've done on my own fork and it seems to work for enabling a local container to backup to a GCloud bucket.

**Comment by CyberDem0n** [comment]
> Would it be acceptable to modify the behavior to always use the AWS bucket for backups if WAL_S3_BUCKET is specified and to always use google cloud storage if WAL_GCS_BUCKET is specified, regardless of the provider?

Sure.

**Comment by michailbrynard** [comment]
I added this update here: https://github.com/zalando/spilo/pull/201/

**Comment by cmeury** [comment]
Heads up for anyone stumbling across the githubs: The `WAL_S3_BUCKET` variable needs to follow a particular naming scheme. The `configure_spilo.py` script extracts the region from it. When it fails to match a region, it falls back to the default value of `placeholders['instance_data']['zone']` minus the last character. This results in the weird hostname you might encounter:

    cat /home/postgres/etc/wal-e.d/env/WALE_S3_ENDPOINT
    https+path://s3-loca.amazonaws.com:443

So, the bucket name must conform to the regular expression `.*(eu-\w+-\d+)-.*`, and the bucket must be located in the region referenced in the name (since spilo constructs the s3 end-point URI using the extracted region). For example: `eu-west-1-my-bucket`

**Comment by cmeury** [comment]
@michailbrynard How are you configuring the AWS credentials? I had to [custom change](https://github.com/ricardo-ch/platform-postgres-spilo/blob/b51e24a06933d8f14a54aba19c01049e9136a527/postgres-appliance/configure_spilo.py#L504) (quick & dirty, need to use the placeholder mechanism) in our fork to allow the credentials to be stored in the `WAL_ENV_DIR` directory.

Also, on another note, I'm worried about how spilo uses envdir: The files are world-readable. But I think it would be an easy PR to just specify UID/GID and permissions when writing the various envdir files.

**Comment by jipperinbham** [comment]
I've been debugging failed backups for a bit and I'm not entirely sure about the `WALE_S3_ENDPOINT` url being crafted. 

According to the [docs](https://docs.aws.amazon.com/general/latest/gr/rande.html#s3_region), the format of the URL is `s3.REGION.amazonaws.com` (i.e. `s3.us-east-1.amazonaws.com`) but the current `configure_spilo.py` is doing `s3-us-east-1.amazonaws.com`.

**Comment by CyberDem0n** [comment]
It looks like `us-east-1` is the only region which doesn't support `s3-REGION.amazonaws.com` format :(

First of all I would like to explain why do we have following code: https://github.com/zalando/spilo/blob/master/postgres-appliance/configure_spilo.py#L496-L500

In the beginning we were doing deployments in eu-west-1, but after some time we figured out that ping between Germany and Ireland is way too high for us and we decided to switch to eu-central-1.
But unfortunately it didn't work reliably if you just created a new S3 bucket in eu-central-1 and after that immediately deployed Spilo. Wal-e was failing to push basebackup and archive wals because it wasn't able to figure out in which regions this new bucket is located. That's why we kind of implemented "auto-detection" of region.

I think that all this magic first of all could be improved a little bit:
* An regexp, detecting a region (if it is a part of bucket name) has to be changed to get rid from "eu-" constraint.
* URL format has to be changed to `s3.REGION.amazonaws.com` in order to be compatible with us-east-1 region

On top of that I think we should make it possible to explicitly specify `WALE_S3_ENDPOINT` as environment variable. And off course, if it is specified, it must take precedence over "auto-detection".

P.S. it seems to be relatively simple pull request and you are welcome to submit it :)

**Comment by ants** [comment]
I actually implemented reading WALE_S3_ENDPOINT from environment just a week ago for an OpenShift deployment. Let me see if I can upstream that change ahead of rest of the changes.
